File: functions/api/api-game.js
function ApiGame(api, database)
{
	this.getGame = function(request, response)
	{
		const sessionToken = request.get(api.SESSION_TOKEN)
		const gameId = request.param('gameId')
	
		database.account.bySessionToken(sessionToken)
		.then(accountDoc =>
		{
			const gameRef = accountDoc.gameRef(gameId)

			if (gameRef)
			{
				database.game.byRef(gameRef)
				.then(gameDoc =>
				{
					database.player.byGameRef(gameRef)
					.then(playerDocs =>
					{
						response
							.status(200)
							.json(gameDoc.json(playerDocs))
					})
				})
			}
			else
			{
				response.status(403).send()
			}
		})
		.catch(error =>
		{
			response.status(401).send()
		})
	}
}

module.exports = ApiGame

File: src/main/kotlin/com/mauriciotogneri/momowars/api/ApiGame.kt
package com.mauriciotogneri.momowars.api

import com.mauriciotogneri.momowars.exception.ForbiddenException
import com.mauriciotogneri.momowars.exception.InternalServerErrorException
import com.mauriciotogneri.momowars.express.Request
import com.mauriciotogneri.momowars.express.Response
import com.mauriciotogneri.momowars.express.headerParam
import com.mauriciotogneri.momowars.express.pathParam

class ApiGame : BaseApi()
{
    fun getGame(request: Request, response: Response)
    {
        process(response) { database ->
            val sessionToken = request.headerParam(Api.SESSION_TOKEN)
            val gameId = request.pathParam("gameId")

            checkNotEmpty(sessionToken)

            val documentAccount = database.account.bySessionToken(sessionToken)

            val gameRef = documentAccount.gameRef(gameId) ?: throw ForbiddenException()

            val documentGame = database.game.byRef(gameRef)

            response.status(OK).json(documentGame.toJson())
        }
    }

    fun createGame(request: Request, response: Response)
    {
        process(response)
        {
            throw InternalServerErrorException()
        }
    }

    fun getOpenGames(request: Request, response: Response)
    {
        process(response) { database ->

            val sessionToken = request.headerParam(Api.SESSION_TOKEN)

            checkNotEmpty(sessionToken)

            database.account.bySessionToken(sessionToken)

            val games = database.game.getOpenGames()
            val list = games.map { it.basicJson() }

            response.status(OK).json(list)
        }
    }
}
